### PR TITLE
143 unrelevant devices get deleted by cb delete entity

### PR DESCRIPTION
**Pull request by djs0109**
Since the IoTAgent API does not provide filter functionality, I implement this in the IoTAgentClient of FiLiP. In this way the function get_device_list( ) can now return the devices filtered by entity_id.

### CODE REVIEW

**Review by tstorek — CHANGES_REQUESTED** [review]
Thanks for the contribution. only small remakrs :)

File: filip/clients/ngsi_v2/iota.py [hunk 1]
```diff
@@ -341,12 +339,37 @@ def get_device_list(self, *, limit: int = None, offset: int = None,
...
+                    # filter with entity id
+                    devices = self.filter_device_list(devices, entity_name=[entity])
+                return devices
             res.raise_for_status()
         except requests.RequestException as err:
             self.log_error(err=err, msg=None)
             raise
 
+    @staticmethod
+    def filter_device_list(devices: List[Device],
```
[the remark this answers is not in the document]
Comment (djs0109) [reply filip/clients/ngsi_v2/iota.py:353]: Yes good point, then this function can also be invoked more simply by other modules

File: tests/clients/test_ngsi_v2_cb.py [hunk 2]
```diff
@@ -773,6 +776,44 @@ def test_patch_entity(self) -> None:
...
+                device_id=device_id,
+                entity_type=entity_type,
+                entity_name=entity_id
+            )
+            devices.append(device)
+        self.iotac.post_devices(devices=devices)
+
+        # delete an entity and the related device one by one
+        n_devices = 20
+        while True:
```
[the remark this answers is not in the document]
Comment (djs0109) [reply tests/clients/test_ngsi_v2_cb.py]: Done. 926ee6ef6a32a045a0ca684e595ab3b3d9531f35

File: tests/clients/test_ngsi_v2_cb.py [hunk 3]
```diff
@@ -773,6 +776,44 @@ def test_patch_entity(self) -> None:
...
+                service_path=settings.FIWARE_SERVICEPATH,
+                device_id=device_id,
+                entity_type=entity_type,
+                entity_name=entity_id
+            )
+            devices.append(device)
+        self.iotac.post_devices(devices=devices)
+
+        # delete an entity and the related device one by one
+        n_devices = 20
```
[the remark this answers is not in the document]
Comment (djs0109) [reply tests/clients/test_ngsi_v2_cb.py]: see above.

File: tests/clients/test_ngsi_v2_cb.py [hunk 4]
```diff
@@ -773,6 +776,44 @@ def test_patch_entity(self) -> None:
...
+        while True:
+            if not devices:
+                break
+            device = devices.pop()
+            n_devices += -1
```
[the remark this answers is not in the document]
Comment (djs0109) [reply tests/clients/test_ngsi_v2_cb.py]: see above.

File: tests/clients/test_ngsi_v2_cb.py [hunk 5]
```diff
@@ -773,6 +776,44 @@ def test_patch_entity(self) -> None:
...
+            self.client.delete_entity(entity_id=device.entity_name, entity_type=device.entity_type, delete_devices=True)
+            self.assertEqual(len(self.iotac.get_device_list()), n_devices)
```
[the remark this answers is not in the document]
Comment (djs0109) [reply tests/clients/test_ngsi_v2_cb.py]: see above.

**Comment by tstorek** [comment]
@djs0109 I made some minor modifications and added some more tests. Please double-check if everything works as expected.

**Comment by djs0109** [comment]
LGTM

**Comment by tstorek** [comment]
@djs0109 thank you very much!

**Pull request merged**